app/core/downloader/ytdlp.py

```
@@ -0,0 +1,113 @@
+        ydl_params = cast("yt_dlp.YoutubeDL.Params", ydl_opts)  # type: ignore[arg-type]
+        with yt_dlp.YoutubeDL(ydl_params) as ydl:
+            info = ydl.extract_info(direct_url, download=True)
+            if info is None:
+                logger.error("yt-dlp did not return info dict.")
+                raise DownloadError("yt-dlp did not return info dict.")
+            filename = ydl.prepare_filename(info)
+            logger.success("Download finished: %s", filename)
+            return Path(filename), cast(Dict[str, Any], info)
+    except YTDLPDownloadError as exc:
+        logger.error("yt-dlp download failed: %s", exc)
+        raise DownloadError(str(exc))
```

Ruff (0.14.10)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)